### PR TITLE
feat(bottlerocket): support multiple cluster-dns-ip configuration

### LINKED ISSUE #4836: bottlerocket with mulitple cluster-dns
### Description

**Observed Behavior**:

Not able to set mulitple ips for cluster-dns-ip

bottlerocket allows you to set it via

```
# Valid, single IP
[settings.kubernetes]
"cluster-dns-ip" = "10.0.0.1"

# Also valid, multiple nameserver IPs
[settings.kubernetes]
"cluster-dns-ip" = ["10.0.0.1", "10.0.0.2"]
```

Trying to set it like

```
[settings.kubernetes]
"cluster-dns-ip" = ["10.0.0.1", "10.0.0.2"]
```

Gives the following log

```
﻿karpenter-f7cdc49f4-25kbx controller 2023-10-15T02:10:53.850Z	ERROR	controller	Reconciler error	{"commit": "322822a", "controller": "machine.lifecycle", "controllerGroup": "karpenter.sh", "controllerKind": "Machine", "Machine": {"name":"workers-gmsx5"}, "namespace": "", "name": "workers-gmsx5", "reconcileID": "8ace2c6e-45f7-4bb4-84d8-79388dde83b0", "error": "launching machine, creating instance, getting launch template configs, getting launch templates, creating launch template, invalid UserData toml: cannot decode TOML array into struct field bootstrap.BottlerocketKubernetes.ClusterDNSIP of type *string"}
```

Looking at the code it takes a string but even if you do it with a comma it always just has one which is the default kube-dns

https://github.com/aws/karpenter/blob/main/pkg/providers/amifamily/bootstrap/bottlerocket.go#L59

**Expected Behavior**:

be able to have a /etc/resolv.conf in pods like

node-local helps to stop against loading a lot of coredns servers so we aren't rate limited by the vpc resolver. 

```
$ cat /etc/resolv.conf
search testing.svc.cluster.local svc.cluster.local cluster.local ec2.internal
nameserver 169.254.20.10
nameserver 172.20.0.10
```

**Reproduction Steps** (Please include YAML):

Tried both of the following

```
     userData: |
        [settings.kubernetes]
        "cluster-dns-ip" = ["169.254.20.10","172.20.0.10"]
```

and

```
     userData: |
        [settings.kubernetes]
        "cluster-dns-ip" = "169.254.20.10,172.20.0.10"
```

**Versions**:
- Chart Version: 
- Kubernetes Version (`kubectl version`): 

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by shaikatzz**
## Summary

This PR implements support for multiple DNS IPs in Bottlerocket cluster-dns-ip configuration, fixing the limitation where only the first DNS IP was used from KubeletConfig.ClusterDNS array.

## Problem (Closes #4836)

Issue #4836 reported that Bottlerocket was unable to handle multiple DNS IPs in the cluster-dns-ip configuration. The core issues were:

1. **Only First IP Used**: The code was taking only the first DNS IP: `s.Settings.Kubernetes.ClusterDNSIP = &b.KubeletConfig.ClusterDNS[0]`
2. **Type Mismatch**: The `ClusterDNSIP` field was defined as `*string` but users needed to pass arrays like `["10.0.0.1", "10.0.0.2"]`
3. **TOML Decoding Error**: Attempting to set multiple IPs resulted in: `cannot decode TOML array into struct field bootstrap.BottlerocketKubernetes.ClusterDNSIP of type *string`

### Core Changes:
1. **Changed field type**: `ClusterDNSIP` from `*string` to `[]string` in BottlerocketKubernetes struct
2. **Fixed assignment logic**: Changed from `&b.KubeletConfig.ClusterDNS[0]` to `b.KubeletConfig.ClusterDNS` to use all DNS IPs
3. **Added custom TOML handling**: Custom UnmarshalTOML logic to handle both string and array formats in user data
4. **Improved TOML output**: Single IPs output as strings, multiple IPs as arrays

### Key Code Changes:

**Before:**
```go
// Only used first DNS IP
s.Settings.Kubernetes.ClusterDNSIP = &b.KubeletConfig.ClusterDNS[0]

// Field was *string
ClusterDNSIP *string `toml:"cluster-dns-ip,omitempty"`
```

**After:**
```go
// Uses all DNS IPs
s.Settings.Kubernetes.ClusterDNSIP = b.KubeletConfig.ClusterDNS

// Field is []string
ClusterDNSIP []string `toml:"cluster-dns-ip,omitempty"`
```

### Files Modified:
- `pkg/providers/amifamily/bootstrap/bottlerocket.go` - Fixed to use full ClusterDNS array instead of just first element
- `pkg/providers/amifamily/bootstrap/bottlerocketsettings.go` - Changed field type and added custom TOML handling
- `pkg/providers/amifamily/bootstrap/bottlerocket_test.go` - Added comprehensive unit tests (298 lines)
- `pkg/providers/launchtemplate/suite_test.go` - Added integration tests
- Added 4 new golden test data files for various scenarios

### Key Features:
- ✅ **Multiple DNS IP Support**: Now properly handles all DNS IPs from KubeletConfig.ClusterDNS
- ✅ **Backward Compatibility**: Single IP strings are automatically converted to arrays internally
- ✅ **Custom User Data Support**: Handles both string and array formats in user-provided TOML
- ✅ **Optimal TOML Output**: Single IPs as strings, multiple IPs as arrays

## Testing

- **16 Bottlerocket bootstrap tests** passing
- **13 new unit tests** covering array/string handling, edge cases, and override behavior
- **Integration tests** in launch template suite
- **Golden test data** for various input/output scenarios
- **Backward compatibility** verified

### KubeletConfig Usage (Now Works):
```yaml
apiVersion: karpenter.k8s.aws/v1
kind: EC2NodeClass
spec:
  kubelet:
    clusterDNS:
      - "169.254.20.10"  # node-local-dns
      - "172.20.0.10"    # coredns
```

### Custom User Data (Now Works):
```toml
# Single IP - outputs as string
[settings.kubernetes]
cluster-dns-ip = '169.254.20.10'

# Multiple IPs - outputs as array  
[settings.kubernetes]
cluster-dns-ip = ['169.254.20.10', '172.20.0.10']
```

## Why This Closes #4836

This PR directly addresses all the issues mentioned in #4836:

1. **Fixes "Only First IP" Issue**: Changed from `ClusterDNS[0]` to using the full `ClusterDNS` array
2. **Fixes TOML Decoding Error**: Changed field type from `*string` to `[]string` to properly handle arrays
3. **Enables Multiple DNS IPs**: Users can now set multiple nameserver IPs as requested
4. **Supports node-local-dns**: Enables the exact use case mentioned (169.254.20.10 + 172.20.0.10)

The implementation allows users to achieve the desired `/etc/resolv.conf` configuration with multiple nameservers, solving the rate limiting issues with VPC resolver that the issue author mentioned.

## Backward Compatibility

This change is fully backward compatible:
- Existing single IP configurations continue to work
- Custom user data with single IP strings is automatically handled
- No breaking changes to existing APIs or configurations

### CODE REVIEW

**Comment by demmonico** [comment]
It's a super helpful MR that fixes an old, annoying issue. Are there any plans to merge it soon?

File: pkg/providers/amifamily/bootstrap/bottlerocketsettings.go [hunk 1]
```diff
@@ -118,28 +118,73 @@ type BootstrapCommand struct {
...
-	// unmarshal untyped settings
-	if err := toml.Unmarshal(data, c); err != nil {
+	if err := toml.Unmarshal(processedData, c); err != nil {
 		return err
 	}
 	c.Settings = s.Settings
 	return nil
 }
 
+func (c *BottlerocketConfig) preprocessTOML(data []byte) ([]byte, error) {
```
Comment (jigisha620) [inline pkg/providers/amifamily/bootstrap/bottlerocketsettings.go:139]: The TOML round-tripping here seems heavier than needed. It does full unmarshal then modify followed by re-marshal of the entire userData to convert one field's type on input. A simpler option could be - try unmarshaling as-is first, and if cluster-dns-ip fails as `[]string`, fall back to a temporary struct with `*string` and convert.

**Review by jigisha620 — COMMENTED** [review]
Array cluster-dns-ip was added in Bottlerocket v1.8.0 (https://github.com/bottlerocket-os/bottlerocket/pull/2176). Karpenter already resolves the Bottlerocket AMI version via `resolveAMIVersion()` and uses it for feature gating (see SupportsDefaultBind). The same pattern could be used here I think. If the AMI supports it, always output cluster-dns-ip as an array. When it doesn't, fall back to the first IP as a string (current behavior). This would eliminate `formatKubernetesSettings` entirely and its TOML round-tripping.
